piq/pieapp.py

```diff
@@ -0,0 +1,205 @@
+        self.conv9 = nn.Conv2d(features * 4, features * 4, kernel_size=3, padding=1)
+        self.conv10 = nn.Conv2d(features * 4, features * 8, kernel_size=3, padding=1)
+        self.conv11 = nn.Conv2d(features * 8, features * 8, kernel_size=3, padding=1)
+
+        self.fc1_score = nn.Linear(in_features=120832, out_features=512, bias=True)
+        self.fc2_score = nn.Linear(in_features=512, out_features=1, bias=True)
+        self.fc1_weight = nn.Linear(in_features=2048, out_features=512)
+        self.fc2_weight = nn.Linear(in_features=512, out_features=1, bias=True)
+        self.ref_score_subtract = nn.Linear(in_features=1, out_features=1, bias=True)
+
+        # Term for numerical stability
+        self.EPS = 1e-6
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""
+        Forward pass batch of square patches with shape  (N, C, features, features)
+        Returns:
+            
+        """
+        assert x.shape[2] == x.shape[3] == self.features, \
+            f"Expected square input with shape {self.features, self.features}"
+
+        # conv1 -> relu -> conv2 -> relu -> pool -> conv3 -> relu
+        x3 = F.relu(self.conv3(self.pool(F.relu(self.conv2(F.relu(self.conv1(x)))))))
+        # conv4 -> relu -> pool -> conv5 -> relu
+        x5 = F.relu(self.conv5(self.pool(F.relu(self.conv4(x3)))))
+        # conv6 -> relu -> pool -> conv7 -> relu
+        x7 = F.relu(self.conv7(self.pool(F.relu(self.conv6(x5)))))
+        # conv8 -> relu -> pool -> conv9 -> relu
+        x9 = F.relu(self.conv9(self.pool(F.relu(self.conv8(x7)))))
+        # conv10 -> relu -> pool1-> conv11 -> relU
+        x11 = self.flatten(F.relu(self.conv11(self.pool(F.relu(self.conv10(x9))))))
+        # flatten and concatenate
+        features = torch.cat((self.flatten(x3), self.flatten(x5), self.flatten(x7), self.flatten(x9), x11), dim=1)
+        return features, x11
+
+    def compute_difference(self, features_diff, weights_diff):
+        r"""
+        Args:
+            features_diff: Tensor of shape (N * NUM_PATCHES, C_1)
+            features_diff: Tensor of shape (N * NUM_PATCHES, C_2)
+        Returns:
+            distances
+            weights
+        """
+        # Get scores: fc1_score -> relu -> fc2_score
+        # 0.01 is the sigmoid coefficient
+        distances = self.ref_score_subtract(0.01 * self.fc2_score(F.relu(self.fc1_score(features_diff))))
+
+        weights = self.fc2_weight(F.relu(self.fc1_weight(weights_diff))) + self.EPS
+        return distances, weights
+
+
+class PieAPP(_Loss):
+    r"""
+    Implementation of Perceptual Image-Error Assessment through Pairwise Preference.
+    
+    Expects input to be in range [0, 1] with no normalization and RGB channel order.
+    Input images are croped into smaller patches and final score is mean of patch scores.
+    
+
+    Args:
+        reduction: Reduction over samples in batch: "mean"|"sum"|"none".
```

Comment:
If `final score is mean of patch scores`, then I think `reduction` parameters is a bit redundant. We either should avoid explicit description in the doctoring (which would be weird if this is a part of an algorithm for the metric) or remove other redundancy options. What do you think?

Reply:
Updated docstring description. Final score for _individual_ image  is mean of it's patch scores. So we still may want to reduce results over batch.